### PR TITLE
Mypy for surrogates

### PR DESCRIPTION
**Pull request by AdrianSosic**
This PR finally enables mypy for the surrogates package

### CODE REVIEW

File: baybe/surrogates/base.py [hunk 1]
```diff
@@ -85,7 +89,7 @@ def to_botorch(self) -> Model:
 
 
 @define
-class Surrogate(ABC, SurrogateProtocol, SerialMixin):
+class Surrogate(ABC, SurrogateProtocol, SerialMixin, Generic[_ModelContext]):
```
Comment (AVHopp) [inline baybe/surrogates/base.py:92]: Why is this necessary? It feels weird that the surrogate inherits from what feels like a bare necessity that is required to make mypy happy.

File: baybe/surrogates/custom.py [hunk 2]
```diff
@@ -96,7 +96,7 @@ def _estimate_moments(self, candidates: Tensor, /) -> tuple[Tensor, Tensor]:
             torch.from_numpy(results[1]).pow(2).to(DTypeFloatTorch),
         )
 
-    def _fit(self, searchspace: SearchSpace, train_x: Tensor, train_y: Tensor) -> None:
+    def _fit(self, train_x: Tensor, train_y: Tensor, context: Any) -> None:
```
Comment (AVHopp) [inline baybe/surrogates/custom.py:99]: Was this change overlooked previously? Does not seem to be purely relateed to mypy issues

File: baybe/surrogates/ngboost.py [hunk 3]
```diff
@@ -60,12 +61,23 @@ def _make_parameter_scaler_factory(
         # See base class.
 
         # Tree-like models do not require any input scaling
-        return
+        return None
+
+    @staticmethod
```
Comment (AVHopp) [inline baybe/surrogates/ngboost.py:66]: I guess re-introducing is necessary because this is now `static`? Because the function looks identical to the one from the parent class

**Review by AVHopp — COMMENTED** [review]
Some minor things that I do not understand, pls make clear

File: baybe/surrogates/base.py (hunk 1 above)
Comment (AdrianSosic) [reply baybe/surrogates/base.py:92]: Not that this is **not** inheritance, it's purely a typing thing and the syntax required to declare generic classes in python until python 3.12 (in newer versions, this looks different). See:
https://mypy.readthedocs.io/en/stable/generics.html#defining-generic-classes

It lets us relate the output of `_get_model_context` to the input of `_fit` and thus allows the type checker to do its job properly. 

If more questions / still unhappy, let me know. Otherwise, resolve

File: baybe/surrogates/ngboost.py (hunk 3 above)
Comment (AdrianSosic) [reply baybe/surrogates/ngboost.py:66]: what do you mean with "re-introduce"? Perhaps you overlooked: it overrides the return to `None`

File: baybe/surrogates/base.py (hunk 1 above)
Comment (AVHopp) [reply baybe/surrogates/base.py:92]: Ok, but why is it generic here? Is this simply because `_ModelContext` can basically be `Any`, and this is just the proper way to encode that?

File: baybe/surrogates/ngboost.py (hunk 3 above)
Comment (AVHopp) [reply baybe/surrogates/ngboost.py:66]: OK, then just overlooked that or checked at the wrong place.

File: baybe/surrogates/base.py (hunk 1 above)
Comment (AVHopp) [reply baybe/surrogates/base.py:92]: Makes sense :)
